Add extended cvmfs stat call to libcvmfs

nhazekam opened the pull request:
This addition to libcvmfs provides a functions for retrieving not only the stat information, but also information specific to CVMFS like checksum and the stored names. This also adds tests for stat and ext_stat in t_libcvmfs.

Note: this coincides with #2155 so one will need to be rebased off the other prior to merging.

siscia commented:
I find curious that you opted for `char*` instead of `std::string`, is there any technical reason?

nhazekam commented:
If you are referring inside of the `cvmfs_stat` structure, I am using chars so that this is accessible in c, not just c++.  The exceptions are for checksum and xattr so that the CVMFS structured data remains. If a c application were interested in this data, there would likely have to be a conversion or helper function to access the data.

nhazekam commented:
@jblomer Can you take a look when you get the chance.

jblomer commented on cvmfs/directory_entry.h:
```diff
@@ -195,6 +196,31 @@ class DirectoryEntryBase {
+   * information such as hash.
+   * @return the struct cvmfs_stat for this DirectoryEntry
+   */
+  inline void GetCVMFSStatStructure(struct cvmfs_attr *attr) const {
```
I think we should move this to libcvmfs, and convert from the other side, like `cvmfs_attr::FromDirent`

jblomer commented on cvmfs/directory_entry.h:
```diff
@@ -18,6 +18,7 @@
 #include "bigvector.h"
 #include "compression.h"
 #include "hash.h"
+#include "libcvmfs.h"
```
It might work fine but I'd prefer not to include this header in the internal code.

jblomer commented on cvmfs/directory_entry.h:
```diff
@@ -195,6 +196,31 @@ class DirectoryEntryBase {
+    attr->st_rdev  = rdev();
+    attr->st_size  = size();
+    attr->mtime    = mtime_;
+    attr->st_blksize = 4096;  // will be ignored by Fuse
```
Perhaps we can skip this and the next entry, as they are anyway meaningless

jblomer commented on test/common/catalog_test_tools.h:
```diff
@@ -127,6 +126,7 @@ class CatalogTestTool : public ServerTool {
   bool Apply(const std::string& id, const DirSpec& spec);
   bool ApplyAtRootHash(const shash::Any& root_hash, const DirSpec& spec);
   bool AddNestedCatalog(const shash::Any& root_hash, const std::string& path);
+  bool FindEntry(const shash::Any& root_hash, const std::string& path, catalog::DirectoryEntry *entry);
```
Line break to stay within the 80 chars/line limit.

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -63,10 +69,16 @@ class T_Libcvmfs : public ::testing::Test {
   string opt_cache_;
   FILE *fdevnull_;
+
+  int fd_cwd_;
```
This and the following member are unused, I think.

jblomer commented on test/unittests/t_libcvmfs.cc:
```diff
@@ -264,3 +278,164 @@ TEST_F(T_Libcvmfs, Templating) {
+  EXPECT_EQ(-1, retval);
+  cvmfs_attr_free(attr);
+
+  /* Finalize and close repo and options */
```
Missing `cvmfs_detach_repo()`

nhazekam commented:
@jblomer I think this is ready for another look. This was also rebased on the recent changes and addresses a similar issue in the test. I also moved the version and struct size to the constructor function for `cvmfs_attr`. If this should still be in the setting function please let me know.

nhazekam commented:
@jblomer I updated the calls you mentioned. Also, I agree and updated `cvmfs_stat_attr` so that it looks at the link information. I checked that the test would fail if the call resolved the link, but if there are additional checks you would like me to add please let me know.
